src/lib/optimizer/optimizer.cpp

```diff
@@ -213,16 +213,22 @@ std::shared_ptr<Optimizer> Optimizer::create_default_optimizer() {
 
   optimizer->add_rule(std::make_unique<JoinToPredicateRewriteRule>());
 
-  // The SemiJoinReductionRule is very sensitive to the predicate placement and order present when it is applied. In
-  // general, running the PredicatePlacementRule and the PredicateReorderingRule before the SemiJoinReductionRule is
-  // beneficial. However, TPC-H Q 21 (that is already long-running) degrades drastically. See:
-  // https://github.com/hyrise/hyrise/pull/2536#issuecomment-1423076256
-  // TODO(anyone): Re-evaluate this in the future.
+  // Run the PredicatePlacementRule a second time so that semi/anti joins created by the SubqueryToJoinRule, the
+  // JoinToSemiJoinRule, or predicates created by the JoinToPredicateRewriteRule are properly placed, too. Also run the
+  // PredicateReorderingRule before the SemiJoinReductionRule to order semi/anti joins before we add semi join
+  // reductions. Otherwise, we might add unnecessary reductions.
+  // Note: Pushing all newly created predicates down is not always beneficial. Examples are the TPC-DS queries 13 and
+  // 48. Here, we join a table with a foreign-key relationship and no additional filters on that table without using it
+  // further. The superfluous semi-join is pushed down, degrading performance.
+  // TODO(dey4ss): Revisit and improve this once Hyrise really supports foreign keys/INDs and all FKs are set - either
```

Comment:
I would write out INDs and FKs in this comment and briefly tell what join avoidance is in 5 words.

Reply:
Looked into the plans of degraded queries and removed the entire part since it is not a problem anymore. However, I will document what else causes problems in the PredicateReorderingRule header